Implement recognition of qualified names in CB input

vitvakatu opened the pull request:
### Pull Request Description

Closes #5356 

Implements recognition of qualified names in the input of the CB. Now accepting of the suggestion produces a valid code, and only necessary imports are added.


https://user-images.githubusercontent.com/6566674/225487445-855dfb04-47c7-44c9-bc92-4bb5663d500a.mp4


https://user-images.githubusercontent.com/6566674/225487486-028d1ae0-02e0-4bbc-a623-7d4e620c87d9.mp4



### Important Notes

- Filtering by inserted code no longer works, but instead, we filter out entries with qualified names not matching the edited accessor chain. It may result in worsened behavior when typing expressions with local variables as `this` argument.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

vitvakatu committed:
Implement recognition of qualified names in CB input

farmaazon commented on app/gui/src/controller/searcher.rs:
```diff
@@ -2054,4 +2112,88 @@ pub mod test {
+                input:            "Foo.Gee.".to_string(),
+                expected_code:    "operator1 = Bar.bar_method".to_string(),
+                expected_imports: vec!["local.Project.Foo.Bar".to_string()],
+            },
```
Please add a case of input being a full qn path (e.g. "local.Project.Foo.Bar." and bar_method entry).

vitvakatu replied:
Good catch, it highlighted a bug. Fixed now.

farmaazon commented on app/gui/src/controller/searcher/input.rs:
```diff
@@ -371,16 +380,117 @@ impl InsertedSuggestion {
+            let every_operand_is_name = context
+                .enumerate_operands()
+                .flatten()
+                .all(|opr| matches!(opr.item.arg.shape(), ast::Shape::Cons(_)));
```
Technically the Qualified Names could be also lowercase, what resolves to `ast::Shape::Var` variant (most common example is `local` namespace, but modules theoretically could be lowercase as well AFAIK).

vitvakatu replied:
👍🏻 Fixed.

vitvakatu committed:
Support `local.Project.` qualified names.

wdanilo commented on app/gui/src/controller/searcher.rs:
```diff
@@ -154,6 +154,24 @@ impl Default for Actions {
+pub enum RequiredImport {
+    /// No import needed.
+    #[default]
+    None,
```
Why are we not reusing `Option` here? Option has so many utils defined that in (almost) all such cases its just a better option :) Also, it would make some code easier to write, like the one `.flat_map(|requirement| match requirement [...]` below.

vitvakatu replied:
I don't have a strong preference here. The code would be a bit longer in signatures, but we can save some space elsewhere. I'll change it.

vitvakatu committed:
Merge branch 'wip/farmaazon/cursor-aware-cb-183521918' into wip/vitvakatu/recognize-qualified-names

vitvakatu committed:
Use Option<RequiredImport>

vitvakatu committed:
Disable tests broken because there is no MethodCallInfo anymore.

farmaazon commented:
## First round of QA

1. Missing import for project name (`Base` in the case below)

![Screenshot from 2023-03-17 11-15-11](https://user-images.githubusercontent.com/3919101/225878828-a3133c26-be1e-4556-a31e-0db0e60296e4.png)

2. UX: When trying to put `Table.new` CB assumed the longest possible QN, I think it should rather assume the shortest (especially because in this case the shortest is already imported):

![Screenshot from 2023-03-17 11-07-45](https://user-images.githubusercontent.com/3919101/225880242-2a4b1ca6-ec6e-49da-a6d4-af7593fa859a.png)

3. Also, there is a regression, but caused by design, not by code: Before typing `Table.` showed me only `Table` methods, because `Table.` was a pattern, and it was matched against the code inserted by the suggestion (like `Table.new`). 

Unfortunately, even #5232 seem to not fully address this issue.

Not sure about the solution. Perhaps a quick fix could be additional filtering out entries whose FQN does not match the current "context".

vitvakatu commented:
> Missing import for project name (Base in the case below)

Interesting, it should work, and I can even add a passing unit test for that. Investigating.

> UX: When trying to put Table.new CB assumed the longest possible QN, I think it should rather assume the shortest (especially because in this case the shortest is already imported):

I will investigate.

vitvakatu committed:
Import project's main module

vitvakatu committed:
Merge branch 'wip/farmaazon/cursor-aware-cb-183521918' into wip/vitvakatu/recognize-qualified-names

vitvakatu committed:
Consider the shortest possible name first

vitvakatu committed:
Filter entries by context

vitvakatu commented:
@farmaazon spotted issues were addressed.

> Missing import for project name (Base in the case below)

Fixed, though I don't like the Engine behavior in some cases. See https://github.com/enso-org/enso/discussions/6026

> UX: When trying to put Table.new CB assumed the longest possible QN, I think it should rather assume the shortest (especially because in this case the shortest is already imported):

Nice catch, fixed.

> Not sure about the solution. Perhaps a quick fix could be additional filtering out entries whose FQN does not match the current "context".

I implemented it (in a separate branch), but I don't think it is what we really want. `context` is not only a QN, but also `this` argument, and it means we will effectively show no entries for inputs like `operator1.`. Perhaps we can detect QN only, but still feels hacky. What do you think?

Also, I left one TODO in the code for unit tests – it needs refactoring, I will do it later this day. It shouldn't block further QA.


https://user-images.githubusercontent.com/6566674/226603289-66a395bd-bdf6-450f-be16-50dfca2a03ac.mp4






<blockquote><img src="https://opengraph.githubassets.com/179693744d3375f9a67f47b3f1fe407480ea2159fb8fd4bbb5bb33318000885e/orgs/enso-org/discussions/6026" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/orgs/enso-org/discussions/6026">Inconsistent names resolution · enso-org/enso · Discussion #6026</a></strong></div><div>Hi! I'm developing recognition of qualified names in CB and faced a weird behavior of the engine. I have a project named Unnamed_14 with the submodule Foo with the method foo. I also have func1 def...</div></blockquote>

vitvakatu committed:
Refactor filtering out by context

vitvakatu committed:
Fix generation of code

vitvakatu committed:
enso-formatter pass

vitvakatu committed:
Linting

vitvakatu committed:
Fix generating code again

vitvakatu commented:
> I implemented it (in a separate branch), but I don't think it is what we really want. context is not only a QN, but also this argument, and it means we will effectively show no entries for inputs like operator1.. Perhaps we can detect QN only, but still feels hacky. What do you think?

After a call with @farmaazon we decided to go with this solution. Now we filter out entries simply by the string representation of `context`. It does not work for local variables well, but we will improve it in the future.

vitvakatu committed:
Fix filtering

farmaazon commented:
## Second round of QA

The previous issues are fixed, but the filtering seems to not working on some configurations; here I see no suggestions of   `Aggregate_Column` constructors, if I'm in the array.

![Screenshot from 2023-03-22 10-48-12](https://user-images.githubusercontent.com/3919101/226890149-e7d95b4f-11b0-43e7-a416-298f1324bae4.png)

But when I just open CB and type `Aggregate_Column.`, it works.

vitvakatu commented:
> The previous issues are fixed, but the filtering seems to not working on some configurations; here I see no suggestions of Aggregate_Column constructors, if I'm in the array.

This is an issue on the parser side, reported https://github.com/enso-org/enso/issues/6041
Using expressions `[Aggregate_Column.]` or `[ Aggregate_Column.` works as desired.

farmaazon commented:
**QA green** No other issues found.

vitvakatu committed:
Use mock_namespace.MockProject in unit tests

vitvakatu committed:
Lint

vitvakatu committed:
Remove TODO
